b288702/verify_contrib.sh

```diff
@@ -81,35 +75,72 @@ if [[ ! -f './b288702.b2sums' ]]; then
     exit 1
 fi
 
-url_base='https://trusted-setup.s3.amazonaws.com/phase2-mainnet'
-
-# Verify phase2 contributions.
+# Verify Phase 2 contributions.
 log "verifying contribution: ${contrib}"
 
 prev=$((contrib - 1))
 prev_file="${proof}_poseidon_${sector_size}gib_b288702_${prev}_small_raw"
-if [[ ! -f ${prev_file} ]]; then
-    log "downloading params: ${file}"
-    curl --progress-bar -O ${url_base}/${prev_file}
-fi
-# Verify checksum even if file was present, in case of incomplete download or corruption.
-# This is especially relevant if another process might have initiated a download now in process.
-log 'verifying downloaded params checksum'
-grep $prev_file b288702.b2sums | b2sum -c
 
+# We should never download parameters for the first contribution (index 0).
+# We generate large params instead.
+if [[ $prev -eq 0 ]]; then
+    if [[ $proof == 'winning' ]]; then
+        phase1_file='phase1radix2m19'
+        phase1_checksum='4a3b6930739967248fee48dbf43e27ee907ab3780132e21d4c7fe37fcebdc87352f1495795178c27799828db9da3696eb6ef19054404b23ec4994883877d96f8'
+    else
+        phase1_file='phase1radix2m27'
+        phase1_checksum='8a5d4e211e3a9817dcdc7d345a25338f261d0b52ab188661dfcb6bada9f2f5ac76925521621d8f87b5b680105973f4b48fb1ec68f65ebf8fdbccbd8d4891e6e9'
+    fi
+
+    if [[ ! -f ${phase1_file} ]]; then
+        error "${phase1_file} is missing. Run: ./download_prereqs_contrib.sh ${proof} ${sector_size} ${contrib}"
+        exit 1
+    fi
+    log 'verifying Phase 1 checksum'
+    echo "${phase1_checksum} ${phase1_file}" | b2sum -c
+
+    # Generate initial phase2 params.
+    initial_large="${proof}_poseidon_${sector_size}gib_b288702_0_large"
+    if [[ ! -f ${initial_large} ]]; then
+        log 'generating initial params'
+        ./phase2 new --${proof} --${sector_size}gib
+
+        # Rename initial params file to replace commit hash at time of
+        # ceremony start with that of the current release (which should be
+        # checked out), so verification will succeed.
+        mv ${proof}_poseidon_${sector_size}gib_$(git rev-parse --short=7 HEAD)_0_large $initial_large
+
+        log "${green}success:${off} finished generating initial phase2 parameters"
+    else
+        log 'use previously generated inital params'
+    fi
+
+    # Verify checksum of generated initial params.
+    log 'verifying initial params checksum'
+    grep $initial_large b288702.b2sums | b2sum -c
+else
+    if [[ ! -f ${prev_file} ]]; then
+        error "${prev_file} is missing. Run: ./download_prereqs_contrib.sh ${proof} ${sector_size} ${contrib}"
+        exit 1
+    fi
+    # Verify checksum even if file was present, in case of incomplete download
+    # or corruption. This is especially relevant if another process might have
+    # initiated a download now in process.
+    log 'verifying downloaded params checksum'
+    grep $prev_file b288702.b2sums | b2sum -c
+fi
```

Comment:
As noted above, this should be its own script the verifier runs exactly once. That script can probably have  a better name the `verify_initial.sh`, though. I think `generate_initial.sh` is probably good.

Reply:
The of having it there is actually that you don't necessarily want to run it. If you split the work and you only want to verify contribution number 5 of a circuit, it would be a waste to create those params.

You could of course not run the `generate_initial.sh` script. But it would make the description way more difficult then just running the generation conditionally based on the contribution you want to verify.

It can well be that I'm missing something and there is a reason to generate them, even if you do not verify the first contribution.

Reply:
I understand the point you're making, and I can probably be convinced. Whichever way we go, *something* must be documented or changed in the README.

Either:
- We make initial parameter-generation explicit, and instruct the verifier to perform it only before verifying the first contribution — something like:
```
./generate_initial.sh && ./verify_contrib.sh winning 32 1
```
Or:
- We add documentation to the effect that verifying contribution 1 is more expensive and is actually performing two distinct steps.

I think it's cleaner and clearer to document the special case by changing the commands as in the first option above. This makes clear the actual process so the verifier understands what is happening. If the verifier *does* understand the process, they should expect to see the initial parameter generation somewhere, and putting the command at top level will make it easier for them to confirm that expectation by inspection.

Reply:
Oh that's easy: I *very much* prefer the first version. It shows the *logical* discrete steps, not necessarily the nitty-gritty, but it gives you an idea of "oh... this thing is quite different".

It's not only about "how long it takes", it is also about "what logs scroll by"

Remember that this reorg started *precisely* because I paid attention to the logs and went "wait... this doesn't... at all agree with what I think I am doing here". Transparency/verbosity in the context of verification is **really** good, as long as it does not become overwhelming.

TLDR: 
`./generate_initial.sh winning 32 && ./verify_contrib.sh winning 32 1`
is **vastly** preferable to
`./verify-stuff-trust-me-it-is-all-automated-mkay winning 32`

I mean heck - I might *want* to disconnect the network cable between the steps, who knows what makes me (the verifier) happy.

Reply:
Thanks, @ribasushi. My opinions was, of course, informed by our discussion of these considerations. Since incorporating your feedback in order to ensure your verification pass provides the assurance we want (of both correctness *and* of confidence in the process), I consider your preference controlling. Let's go with the first option.

Reply:
A slight amendment: we also need to include the download of phase1 parameters (which my example above did not). The existing overloaded download script (with special case for contribution 1) could be used, or we could be even more explicit and break that out into its own script. This leads to either:

```
./download_prereqs_generate_initial.sh winning 32 && ./generate_initial.sh winning 32 && ./verify_contrib.sh winning 32 1
```
Or:
```
./download_prereqs_contrib.sh winning 32 1 && ./generate_initial.sh winning 32 && ./verify_contrib.sh winning 32 1
```

Again, I favor the first option.

Reply:
Per discussion with @ribasushi I am amending the recommended first option to avoid grammatical ambiguity:
```
./download_initial_generation_prereqs.sh winning 32 && ./generate_initial.sh winning 32 && ./verify_contrib.sh winning 32 1
```

Reply:
To make sure we are on the same page. I understand and agree. The other contributions (other then the first one) will still be:

    ./download_prereqs_contrib.sh winning 32 7 && ./verify_contrib.sh winning 32 7

correct?

Reply:
Yes — only the first contribution will change.

Reply:
(For each circuit.)